test/performance/perf_test.lua

```diff
@@ -234,6 +251,200 @@ local select_prepare = function(g)
+            assert(type(opts.limit) == 'number')
+            assert(opts.limit > 0)
+
+            local tuples = {}
+
+            for id, replicaset in pairs(vshard.router.routeall()) do
+                local resp, err = replicaset:call(
+                    '_vshard_select_storage',
+                    { space, key, nil, box.index.GT, opts.limit }
+                )
+                if err ~= nil then
+                    error(err)
+                end
+
+                for _, v in ipairs(resp) do
+                    table.insert(tuples, v)
+                end
+
+            end
+
+            -- Naive merger.
+            local response = { }
+
+            table.sort(tuples, sort)
+
+            for i = 1, opts.limit do
+                response[i] = tuples[i]
+            end
+
+            return response
+        end
+
+        rawset(_G, '_vshard_select_gt', _vshard_select_gt)
+
+
+        local function _vshard_select_secondary(space_name, index_name, key, opts)
+            assert(type(opts.limit) == 'number')
+            assert(opts.limit > 0)
+
+            local storage_response = {}
+
+            for id, replicaset in pairs(vshard.router.routeall()) do
+                local resp, err = replicaset:call(
+                    '_vshard_select_storage',
+                    { space_name, key, index_name, box.index.EQ, opts.limit }
+                )
+                if err ~= nil then
+                    error(err)
+                end
+
+                storage_response[id] = resp
+            end
+
+            -- Naive merger.
+            local response = { }
+
+            local ind = 0
+            for i = 1, opts.limit do
+                for _, tuples in pairs(storage_response) do
+                    if tuples[i] ~= nil then
+                        response[ind] = tuples[i]
+                    end
+                end
+
+                if ind == opts.limit then
+                    break
+                end
+            end
+
+            return response
+        end
+
+        rawset(_G, '_vshard_select_secondary', _vshard_select_secondary)
+
+
+        local function _vshard_select_customer_by_name_and_age(key)
+            local bucket_id = vshard.router.bucket_id_strcrc32(key)
+
+            return vshard.router.callrw(
+                bucket_id,
+                '_vshard_select_customer_by_name_and_age_storage',
+                { key }
+            )
+        end
+
+        rawset(_G, '_vshard_select_customer_by_name_and_age', _vshard_select_customer_by_name_and_age)
+    ]])
+
+    for _, server in ipairs(g.cluster.servers) do
+        server.net_box:eval([[
+            local function _vshard_insert_storage(space_name, tuple, bucket_id)
+                local space = box.space[space_name]
+                assert(space ~= nil)
+
+                assert(space.index.bucket_id ~= nil)
+                tuple[space.index.bucket_id.parts[1].fieldno] = bucket_id
```

Comment:
crud sets bucket id in a tuple on router. Maybe the comparison will be a bit more fair, if we'll do this on router for the vshard case too.

Reply:
Changed